include/SDL3/SDL_ioasync.h

```diff
@@ -0,0 +1,500 @@
+    Uint64 bytes_transferred;  /**< actual number of bytes that were read/written. */
+    void *userdata;    /**< pointer provided by the app when starting the task */
+} SDL_IOAsyncOutcome;
+
+/**
+ * An opaque handle for asynchronous i/o tasks.
+ *
+ * Each asynchronous read or write operation generates a task, which will
+ * complete at some time in the future. This handle is used to track the
+ * progress of that task.
+ *
+ * Tasks can be queried for completion individually, or in a group using
+ * an SDL_IOAsyncGroup.
+ *
+ * \since This struct is available since SDL 3.0.0.
+ *
+ * \sa SDL_ReadIOAsync
+ * \sa SDL_WriteIOAsync
+ */
+typedef struct SDL_IOAsyncTask SDL_IOAsyncTask;
+
+/**
+ * A logical grouping of SDL_IOAsyncTask handles.
+ *
+ * When starting an asynchronous operation, you specify a group for the new
+ * task. A group can be later be asked if any individual task in it has
+ * completed, allowing an app to manage multiple pending tasks in one place,
+ * regardless of the order in which they complete.
+ *
+ * \since This struct is available since SDL 3.0.0.
+ *
+ * \sa SDL_CreateIOAsyncGroup
+ * \sa SDL_ReadIOAsync
+ * \sa SDL_WriteIOAsync
+ * \sa SDL_GetIOAsyncResult
+ * \sa SDL_WaitIOAsyncResult
+ */
+typedef struct SDL_IOAsyncGroup SDL_IOAsyncGroup;
+
+/**
+ * Use this function to create a new SDL_IOAsync object for reading from
+ * and/or writing to a named file.
+ *
+ * The `mode` string understands the following values:
+ *
+ * - "r": Open a file for reading only. It must exist.
+ * - "w": Open a file for writing only. It will create missing files or truncate existing ones.
+ * - "r+": Open a file for update both reading and writing. The file must
+ *   exist.
+ * - "w+": Create an empty file for both reading and writing. If a file with
+ *   the same name already exists its content is erased and the file is
+ *   treated as a new empty file.
+ *
+ * There is no "b" mode, as there is only "binary" style i/o, and no "a" mode
+ * for appending, since you specify the position when starting a task.
+ *
+ * This function supports Unicode filenames, but they must be encoded in UTF-8
+ * format, regardless of the underlying operating system.
+ *
+ * This call is _not_ asynchronous; it will open the file before returning,
+ * under the assumption that doing so is generally a fast operation. Future
+ * reads and writes to the opened file will be async, however.
+ *
+ * \param file a UTF-8 string representing the filename to open.
+ * \param mode an ASCII string representing the mode to be used for opening
+ *             the file.
+ * \returns a pointer to the SDL_IOAsync structure that is created or NULL on
+ *          failure; call SDL_GetError() for more information.
+ *
+ * \since This function is available since SDL 3.0.0.
+ *
+ * \sa SDL_CloseIOAsync
+ * \sa SDL_ReadIOAsync
+ * \sa SDL_WriteIOAsync
+ */
+extern SDL_DECLSPEC SDL_IOAsync * SDLCALL SDL_IOAsyncFromFile(const char *file, const char *mode);
```

Comment:
Do you need to pass the task group in as a parameter? I haven't looked at the IORing API, but it seems like async I/O tasks might not be transferrable between task groups? Or are they?

Reply:
Individual tasks go in one group and then come out completed, but different tasks from the same file can go in different groups.